### ISSUE TITLE
Scanner fails with "Repository provenances with a non-empty VCS path are not supported."

### ISSUE THREAD

**Issue by bennati** [oss-review-toolkit/ort#6070]
Scanning a git-repo project fails with the following error:

```
18:04:28.803 [main] INFO  org.ossreviewtoolkit.scanner.Scanner - Creating file archives for 0 package(s).
18:04:28.803 [main] INFO  org.ossreviewtoolkit.scanner.Scanner - Created file archives for 0 package(s) in 1.84us.
Exception in thread "main" java.lang.IllegalArgumentException: Cannot filter scan results by VCS path that have a repository provenance with a non-blank VCS path because their partial scan result might not contain the path to filter for. The following provenances have a non-blank VCS path: 	RepositoryProvenance(vcsInfo=VcsInfo(type=Git, url=.../protobuf, revision=6484708b736239546051af5ee5713e0116477037, path=upstream/third_party/benchmark), resolvedRevision=6484708b736239546051af5ee5713e0116477037)
	RepositoryProvenance(vcsInfo=VcsInfo(type=Git, url=.../protobuf, revision=6484708b736239546051af5ee5713e0116477037, path=upstream/third_party/googletest), resolvedRevision=6484708b736239546051af5ee5713e0116477037).
	at org.ossreviewtoolkit.scanner.provenance.NestedProvenanceScanResult.filterByVcsPath(NestedProvenanceScanResult.kt:156)
	at org.ossreviewtoolkit.scanner.ScanController.filterByVcsPath(ScanController.kt:316)
	at org.ossreviewtoolkit.scanner.ScanController.getNestedScanResultsByPackage(ScanController.kt:186)
	at org.ossreviewtoolkit.scanner.Scanner.scan(Scanner.kt:158)
	at org.ossreviewtoolkit.scanner.Scanner$scan$4.invokeSuspend(Scanner.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.runScanners(ScannerCommand.kt:228)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.run(ScannerCommand.kt:155)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:198)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:211)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:18)
	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:400)
	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:397)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:415)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:440)
	at org.ossreviewtoolkit.cli.OrtMainKt.main(OrtMain.kt:82)
```

Some warnings are also displayed during the scan execution:

```
WARN  org.ossreviewtoolkit.scanner.Scanner - Could not write scan result for RepositoryProvenance(vcsInfo=VcsInfo(type=Git, url=.../protobuf, revision=6484708b736239546051af5ee5713e0116477037, path=upstream/third_party/googletest), resolvedRevision=6484708b736239546051af5ee5713e0116477037) to ProvenanceBasedPostgresStorage: ScanStorageException: Repository provenances with a non-empty VCS path are not supported.
```

and

```
WARN  org.ossreviewtoolkit.scanner.Scanner - Could not write scan result for RepositoryProvenance(vcsInfo=VcsInfo(type=Git, url=.../protobuf, revision=6484708b736239546051af5ee5713e0116477037, path=upstream/third_party/benchmark), resolvedRevision=6484708b736239546051af5ee5713e0116477037) to ProvenanceBasedPostgresStorage: ScanStorageException: Repository provenances with a non-empty VCS path are not supported.
```

**Comment by sschuberth** [comment]
@bennati I've taken the liberty to edit you post with triple-backticks around code snippets to make it more readable.

**Comment by sschuberth** [comment]
@mnonnenmacher can explain best what's going on here.

**Comment by bennati** [comment]
Hi, any updates @mnonnenmacher ?

**Comment by bennati** [comment]
@mnonnenmacher we tried your suggestion to clear the package-based cache and even disabled the package-based cache entirely but the error is still there.

**Comment by bennati** [comment]
We encountered the same issue also in a git project (not git-repo)

```
________ _____________________
\_____  \\______   \__    ___/ the OSS Review Toolkit, version 1ad5357.
 /   |   \|       _/ |    |
/    |    \    |   \ |    |    Running 'scan' under Java 11.0.16.1 on Linux with
\_______  /____|_  / |____|    8 CPUs and a maximum of 27648 MiB of memory.
        \/       \/
Environment variables:
ORT_CONFIG_DIR = ort-configuration
ORT_DATA_DIR = /builds/oss/oss-review-toolkit/ort-gitlab-ci/.ort
JAVA_HOME = /opt/java/openjdk
ANDROID_HOME = /opt/android-sdk
GOPATH = /tmp/go
Scanning projects with:
	ScanCode (30.1.0)
Scanning packages with:
	ScanCode (30.1.0)
org.ossreviewtoolkit.scanner.ScanStorageException: Repository provenances with a non-empty VCS path are not supported.
	at org.ossreviewtoolkit.scanner.utils.UtilsKt.requireEmptyVcsPath(Utils.kt:31)
	at org.ossreviewtoolkit.scanner.storages.ProvenanceBasedPostgresStorage.read(ProvenanceBasedPostgresStorage.kt:83)
	at org.ossreviewtoolkit.scanner.ProvenanceBasedScanStorageReader$DefaultImpls.read(ScanStorage.kt:84)
	at org.ossreviewtoolkit.scanner.ProvenanceBasedScanStorage$DefaultImpls.read(ScanStorage.kt:143)
	at org.ossreviewtoolkit.scanner.storages.ProvenanceBasedPostgresStorage.read(ProvenanceBasedPostgresStorage.kt:53)
	at org.ossreviewtoolkit.scanner.Scanner.readStoredProvenanceResults(Scanner.kt:468)
	at org.ossreviewtoolkit.scanner.Scanner.readStoredResults(Scanner.kt:416)
	at org.ossreviewtoolkit.scanner.Scanner.scan(Scanner.kt:150)
	at org.ossreviewtoolkit.scanner.Scanner$scan$4.invokeSuspend(Scanner.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.runScanners(ScannerCommand.kt:228)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.run(ScannerCommand.kt:155)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:198)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:211)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:18)
	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:400)
	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:397)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:415)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:440)
	at org.ossreviewtoolkit.cli.OrtMainKt.main(OrtMain.kt:82)
07:09:57.484 [main] WARN  org.ossreviewtoolkit.scanner.Scanner - Could not read scan result for RepositoryProvenance(vcsInfo=VcsInfo(type=Git, url=ssh://git@.../transport-logistics/tour-planning/tour-planning-api, revision=317f844d5a0b0759fd6da18132902a000dc7d5aa, path=tour-planning-api), resolvedRevision=317f844d5a0b0759fd6da18132902a000dc7d5aa) from ProvenanceBasedPostgresStorage: ScanStorageException: Repository provenances with a non-empty VCS path are not supported.
org.ossreviewtoolkit.scanner.ScanStorageException: Repository provenances with a non-empty VCS path are not supported.
	at org.ossreviewtoolkit.scanner.utils.UtilsKt.requireEmptyVcsPath(Utils.kt:31)
	at org.ossreviewtoolkit.scanner.storages.ProvenanceBasedPostgresStorage.write(ProvenanceBasedPostgresStorage.kt:136)
	at org.ossreviewtoolkit.scanner.Scanner.storeProvenanceScanResult(Scanner.kt:547)
	at org.ossreviewtoolkit.scanner.Scanner.runPathScanners(Scanner.kt:376)
	at org.ossreviewtoolkit.scanner.Scanner.scan(Scanner.kt:154)
	at org.ossreviewtoolkit.scanner.Scanner$scan$4.invokeSuspend(Scanner.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.runScanners(ScannerCommand.kt:228)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.run(ScannerCommand.kt:155)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:198)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:211)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:18)
	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:400)
	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:397)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:415)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:440)
	at org.ossreviewtoolkit.cli.OrtMainKt.main(OrtMain.kt:82)
07:10:28.485 [main] WARN  org.ossreviewtoolkit.scanner.Scanner - Could not write scan result for RepositoryProvenance(vcsInfo=VcsInfo(type=Git, url=ssh://git@.../transport-logistics/tour-planning/tour-planning-api, revision=317f844d5a0b0759fd6da18132902a000dc7d5aa, path=tour-planning-api), resolvedRevision=317f844d5a0b0759fd6da18132902a000dc7d5aa) to ProvenanceBasedPostgresStorage: ScanStorageException: Repository provenances with a non-empty VCS path are not supported.
Exception in thread "main" java.lang.IllegalArgumentException: Cannot filter scan results by VCS path that have a repository provenance with a non-blank VCS path because their partial scan result might not contain the path to filter for.
The following provenances have a non-blank VCS path: 	
RepositoryProvenance(vcsInfo=VcsInfo(type=Git, url=ssh://git@.../transport-logistics/tour-planning/tour-planning-api, revision=317f844d5a0b0759fd6da18132902a000dc7d5aa, path=tour-planning-api), resolvedRevision=317f844d5a0b0759fd6da18132902a000dc7d5aa).
at org.ossreviewtoolkit.scanner.provenance.NestedProvenanceScanResult.filterByVcsPath(NestedProvenanceScanResult.kt:156)
	at org.ossreviewtoolkit.scanner.ScanController.filterByVcsPath(ScanController.kt:316)
	at org.ossreviewtoolkit.scanner.ScanController.getNestedScanResultsByPackage(ScanController.kt:186)
	at org.ossreviewtoolkit.scanner.Scanner.scan(Scanner.kt:158)
	at org.ossreviewtoolkit.scanner.Scanner$scan$4.invokeSuspend(Scanner.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.runScanners(ScannerCommand.kt:228)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.run(ScannerCommand.kt:155)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:198)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:211)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:18)
	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:400)
	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:397)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:415)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:440)
	at org.ossreviewtoolkit.cli.OrtMainKt.main(OrtMain.kt:82)
```

**Comment by bennati** [comment]
Reproduced the error also in the file-based cache, after removing the database connection from the config
```
09:19:30.093 [main] WARN  org.ossreviewtoolkit.scanner.Scanner - Could not write scan result for RepositoryProvenance(vcsInfo=VcsInfo(type=Git, url=ssh://git@.../transport-logistics/tour-planning/tour-planning-api, revision=317f844d5a0b0759fd6da18132902a000dc7d5aa, path=tour-planning-api), resolvedRevision=317f844d5a0b0759fd6da18132902a000dc7d5aa) to ProvenanceBasedFileStorage: ScanStorageException: Repository provenances with a non-empty VCS path are not supported.
Exception in thread "main" java.lang.IllegalArgumentException: Cannot filter scan results by VCS path that have a repository provenance with a non-blank VCS path because their partial scan result might not contain the path to filter for. The following provenances have a non-blank VCS path: 	RepositoryProvenance(vcsInfo=VcsInfo(type=Git, url=ssh://git@.../transport-logistics/tour-planning/tour-planning-api, revision=317f844d5a0b0759fd6da18132902a000dc7d5aa, path=tour-planning-api), resolvedRevision=317f844d5a0b0759fd6da18132902a000dc7d5aa).
	at org.ossreviewtoolkit.scanner.provenance.NestedProvenanceScanResult.filterByVcsPath(NestedProvenanceScanResult.kt:156)
	at org.ossreviewtoolkit.scanner.ScanController.filterByVcsPath(ScanController.kt:316)
	at org.ossreviewtoolkit.scanner.ScanController.getNestedScanResultsByPackage(ScanController.kt:186)
	at org.ossreviewtoolkit.scanner.Scanner.scan(Scanner.kt:158)
	at org.ossreviewtoolkit.scanner.Scanner$scan$4.invokeSuspend(Scanner.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.runScanners(ScannerCommand.kt:228)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.run(ScannerCommand.kt:155)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:198)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:211)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:18)
	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:400)
	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:397)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:415)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:440)
	at org.ossreviewtoolkit.cli.OrtMainKt.main(OrtMain.kt:82)
	```

**Comment by bennati** [comment]
We notice this message in many scans, although most are able to complete.

For these scans, the inability to use the cache means the scan time can increase significantly.
Is there any plan to eventually support non-empty provenances?

**Comment by mnonnenmacher** [comment]
@bennati Could you provide a small test project that reproduces this issue? Because so far I have not been able to reproduce it.

**Comment by bennati** [comment]
do you mean reproduce the crash, or the caching issues like: ?
`org.ossreviewtoolkit.scanner.ScanStorageException: Repository provenances with a non-empty VCS path are not supported.`

**Comment by mnonnenmacher** [comment]
@bennati Some project that causes the `IllegalArgumentException` in `filterByVcsPath()`.

**Comment by bennati** [comment]
An update:
After being able to reproduce the issue, I re-scanned the project using a patched ORT that contains the workaround #6166 and the scan completed.
Then, I re-scanned using the unpatched ORT version and the scan completed too, so I suppose there must be some conflict in the cache that caused this issue.
Weird, as disabling the cache was our first unsuccessful attempt.
